Staging/adf4368

skucukar opened the pull request:
The ADF4368 is a high performance, ultra-low jitter, integer-N and fractional-N phase-locked loop (PLL) with integrated VCO ideally suited for frequency conversion applications.

The high performance PLL has a figure of merit of −239 dBc/Hz, very low 1/f noise of normalized −287 dBc/Hz and high PFD frequency that can achieve ultra-low in-band noise and integrated jitter. The ADF4368 can generate any frequency from 800 MHz to 12.8 GHz without an internal doubler, which eliminates the need for sub-harmonic filters. The Σ-Δ modulator includes a 25-bit fixed modulus that allows hertz frequency resolution and an additional 17-bit variable modulus, which allows even finer resolution and flexibility for frequency planning. The 9 dBm output power at 12.8 GHz in single-ended configuration with 16 step power adjust feature makes it very useful for any application.

For multiple frequency conversion applications, such as phase array radar or massive MIMO systems, the outputs of multiple ADF4368 can be aligned by using the SYNC input or EZSync™. The EZSync method is used when it is difficult to distribute the SYNC signal to all devices precisely. For applications that require deterministic delay or delay adjustment capability, a programmable reference to output delay with <1 ps resolution is provided. The reference to output delay is guaranteed across multiple devices and temperature, allowing for predictable and precise multichip alignment.

The simplicity of the ADF4368 block diagram eases development time with a simplified serial-peripheral interface (SPI) register map, external SYNC input, and repeatable multichip phase alignment both in integer mode and fractional mode.

## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus commented on projects/adf4368/src/platform/platform_includes.h:
```diff
@@ -0,0 +1,51 @@
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#ifdef STM32_PLATFORM
```
does the project have stm32 support? if so include it in the builds.json file too.

skucukar replied:
It doesn't support the STM32 platform.

amiclaus replied:
If it doesn't support the STM32 platform then remove all the related additions within code and also the .ioc file.

skucukar replied:
Thank you! Fixed

CiprianRegus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1949 @@
+		return -EINVAL;
+
+	cmd = ADF4368_SPI_WRITE_CMD | reg_addr;
+	if (dev->spi_desc->bit_order) {
+		buff[0] = no_os_bit_swap_constant_8(cmd & 0xFF);
+		buff[1] = no_os_bit_swap_constant_8(cmd >> 8);
+		buff[2] = no_os_bit_swap_constant_8(data);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = data;
+	}
```
I'm a bit confused by this. If the user of the driver is setting the SPI bit order why wouldn't they only use MSB (or what is specified in the datasheet)? In that case you only need to handle one bit order setting.

skucukar replied:
Hi Ciprian, 

I checked this feature. This code snippet is correct but there are other bugs. 
In the "mbed_spi_init" and "no_os_spi_init" functions there is no place to set spi desc->bit_order. Its always 0.
Also in the driver while I am reading I should also swap the bits. I will fix the driver bug.

skucukar replied:
Fixed

CiprianRegus commented on drivers/frequency/adf4368/adf4368.h:
```diff
@@ -0,0 +1,661 @@
+#define ADF4368_POR_DELAY_US			200
+#define ADF4368_LKD_DELAY_MS			9
+
+#define MHZ					MEGA
+#define S_TO_NS					NANO
+#define NS_TO_PS				KHZ_PER_MHZ
+#define PS_TO_NS				KHZ_PER_MHZ
+#define PS_TO_US				MICRO
+#define US_TO_MS				MILLI
+#define MS_TO_S					MILLI
+#define PS_TO_S					PICO
+#define US_TO_FS				NANO
+#define NS_TO_S					NANO
```
These names are generic enough to create future collision with other macros. If you really need to use them, please add the `ADF4368_` prefix.

skucukar replied:
Thank you! Fixed

CiprianRegus commented on drivers/frequency/adf4368/adf4368.h:
```diff
@@ -0,0 +1,661 @@
+ * @struct reg_sequence
+ * @brief ADF4368 register format structure for default values
+ */
+struct reg_sequence {
```
You can make this static. Better yet to move it and the default values to the driver source file. Also, aren't these register values automatically loaded after a reset.

skucukar replied:
This values are the recommended register values for given initial state. After a reset chip's default register values are different.

Struct definition is not static but struct array itself is static in the header file.

I wanted to make the driver similar to adf4382 driver but if its better in the source file I can . 
You mean static struct reg_sequence in the header file and  move the struct array to source file, right?

skucukar replied:
structure and array relocated to source file.

skucukar commented:
V2

- added  "ADF4368_" prefix to calculations 
- modified struct reg_sequence to static struct reg_sequence
- Removed STM32_PLATFORM related additions and .ioc file
- Fix Driver Read - LSB mode

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+/* Charge pump current values expressed in uA */
+static const int adf4368_ci_ua[] = {
+	790,
```
you can shrink this in 2 lines or so.

skucukar replied:
Not changed for readability

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+	if (ret)
+		return ret;
+
+	if (dev->spi_desc->bit_order) {
```
drop brackets.

skucukar replied:
Bracket after "if" is forced with /ci/astyle.sh

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	} while (channel_spacing < ADF4368_CHANNEL_SPACING_MAX);
+
+	if (!en_phase_resync) {
```
drop brackets.

skucukar replied:
Bracket after "if" is forced with /ci/astyle.sh

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+	uint8_t ldwin_pw;
+	uint8_t tmp;
+	int ret;
+
```
drop extra line

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	return 0;
+}
+
```
drop extra line

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	return 0;
+}
+
```
ditto.

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	// Need to set N_INT last to write Double Buffered Registers
+	val = dev->n_int & ADF4368_N_INT_LSB_MSK;
+	ret = adf4368_spi_write(dev, 0x10, val);
```
return directly.

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	// Need to set N_INT last to write Double Buffered Registers
+	val = dev->n_int & ADF4368_N_INT_LSB_MSK;
+	ret = adf4368_spi_write(dev, 0x10, val);
```
return directly.

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	} else {
+		/* Power Down SYNC */
+		ret = adf4368_spi_update_bits(dev, 0x2A, ADF4368_PD_SYNC, 0xff);
```
this return value is lost.

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+
+	return 0;
+}
+
```
drop line

skucukar replied:
Fixed

amiclaus commented on drivers/frequency/adf4368/adf4368.c:
```diff
@@ -0,0 +1,1953 @@
+		// Only Positive Phase
+		if (phase_pol)
+			return -EINVAL;
+		ret = adf4368_set_phase_sdm(dev, phase_fs);
```
return directly

skucukar replied:
Fixed

amiclaus reviewed (commented):
looks good overall. small comments on my side.

skucukar commented:
V3

- removed extra lines
- added direct returns 
- removed brackets for single line if statements

buha commented on drivers/frequency/adf4368/iio_adf4368.c:
```diff
@@ -0,0 +1,1199 @@
+	if (!adf4368)
+		return -EINVAL;
+
+	iio_parse_value(buf, IIO_VAL_INT, &val, NULL);
```
i know this is not going to fail as it is now, but please check the return code of this function everywhere you use it, somebody may add failure points to it in the future and it can silently fail if it's not properly checked

skucukar replied:
Fixed

buha commented on drivers/frequency/adf4368/iio_adf4368.c:
```diff
@@ -0,0 +1,1199 @@
+	{
+		.name = "en",
+		.shared = IIO_SEPARATE,
+		.show = adf4368_iio_read_en,
```
instead of doing 4 different read and 4 different write functions for these attributes you could've done 1 single read and 1 write function

please see how it was done here:
https://github.com/analogdevicesinc/no-OS/blob/cc73a3d70d056d143b783bd846489e2d2819ba7c/drivers/rf-transceiver/hmc630x/iio_hmc630x.c#L326-L366

i know there are lots of places that define a function for an attribute (i've done it too) but we should not be doing it like that anymore, it just takes 200 lines of code to write something that could be done in 50

skucukar replied:
Fixed and tested

buha commented on projects/adf4368/src.mk:
```diff
@@ -0,0 +1,45 @@
+        $(NO-OS)/util/no_os_util.c      \
+        $(NO-OS)/util/no_os_alloc.c
+
+INCS += $(INCLUDE)/no_os_delay.h     \
```
you could use `NO_OS_INCS_DIR += $(INCLUDE)` to replace everything from line 27 to 42

skucukar replied:
Fixed

buha replied:
we discussed in private this is not applicable here because there is a potentially a bug about NO_OS_INC_DIRS on mbed platform

buha reviewed (commented):
solid work, just please see about my other comments and i think it'll be good to merge

skucukar commented:
V4

- Added iio channel attr read-write: IIO Channel attributes combined in to one function 
- Added missing error checks

buha commented on drivers/frequency/adf4368/iio_adf4368.c:
```diff
@@ -0,0 +1,1087 @@
+}
+
+static struct iio_attribute adf4368_iio_attrs[] = {
+	{
```
could you please to the same for device attributes like you did for channel attributes ? 2 functions that manage all these device attributes, it will result in much more compact code.

adf4368_iio_read_device_attr()
adf4368_iio_write_device_attr()

sorry i didn't see this last pass...

buha replied:
and just for completeness, it's not even mandatory to use single function for all, but they just seem so similar that in this case i think it justifies a single function

you could even have a scenario where 1 attribute is so different than the others that you may wanna process it in its own function and have the others be treated in the main function with the big switch case, that's fine too...

skucukar replied:
or I can make separated snprintf for every case

skucukar replied:
Fixed

buha commented on projects/adf4368/src.mk:
```diff
@@ -0,0 +1,47 @@
+        $(NO-OS)/util/no_os_util.c      \
+        $(NO-OS)/util/no_os_alloc.c
+
+INCS += $(NO_OS_INCS_DIR)/no_os_delay.h     \
```
you did this above:
```
NO_OS_INCS_DIR += $(INCLUDE)
```

so delete all thise section, line 29 to 44

skucukar replied:
when I delete them it does not compile

buha commented on projects/adf4368/Makefile:
```diff
@@ -0,0 +1,12 @@
+PLATFORM = mbed
+
+# Select the example you want to enable by chossing y for enabling and n for disabling
+BASIC_EXAMPLE = n
```
(commenting something that i commented on another PR that is being reviewed right now)

i don't think editing Makefiles is good to set the example to build even though we have existing projects done this way... let's strive to make the user interaction easy so that they run one of these commands to choose the example to build:

```
make EXAMPLE=basic_example
make EXAMPLE=iio_example
```

Basically you could default to basic like this:
```
EXAMPLE ?= basic_example
```

And take action based on it:
```
ifeq ($(INTERFACE), iio_example)
...
endif
```

skucukar replied:
If its possible I would like to continue with traditional approach. Our very similar product ADF4382 uses the traditional approach. So we want them to be really similar for the customer experience.

skucukar replied:
Fixed

buha commented on projects/adf4368/README.rst:
```diff
@@ -0,0 +1,428 @@
+
+	# Select the example you want to enable by choosing y for enabling and n for disabling
+	BASIC_EXAMPLE = y
+	IIO_EXAMPLE = n
```
this will need to change when you switch to the `make EXAMPLE=basic_example` approach

skucukar commented:
V5

- Add Device Attributes into one function
- Changed Makefile as describe 
- Bug Fix mbed_spi missing bit_order parameter
- Fix Project README
- Relocated default regs to adf4368.c from adf4368.h

buha commented:
to be merged when CI passes
